redhat-appstudio/remote-secret
# Status condition metric

### What does this PR do?
Provide individual to all remotesecrets metric unveiling information about it is status. Simular to `externalsecret_status_condition` https://external-secrets.io/main/api/metrics/

### Screenshot/screencast of this PR

![Знімок екрана 2024-01-16 о 18 44 06](https://github.com/redhat-appstudio/remote-secret/assets/1614429/03d7b5cf-7960-4e13-8b63-b417467dcd85)


### What issues does this PR fix or reference?
https://issues.redhat.com/browse/SVPI-578

### How to test this PR?
How to get metric
```
 kubectl exec $(kubectl get pods -n remotesecret -l control-plane=controller-manager --no-headers -o custom-columns=":metadata.name") -c manager -n remotesecret --  curl -s  http://localhost:8080/metrics | grep redhat_appstudio_remotesecret_status_condition
```


1.  Create RS without data and no targets
```yaml
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: $OBJ_NAME
spec:
  secret:
    generateName: some-secret-
    labels:
      a: b
    annotations:
      c: d

```
expected metric
```
redhat_appstudio_remotesecret_status_condition{condition="DataObtained",name="test-conditionmetric-l1krtf2wksbmz",namespace="ns-conditionmetric-b8gc9c6trabq9",status="False"} 1
```
2.  Create RS with data and no targets
```yaml
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: $OBJ_NAME
spec:
  secret:
    generateName: secret-from-remote-
stringData:
  a: b
  c: d
```
expected metric
```
redhat_appstudio_remotesecret_status_condition{condition="DataObtained",name="test-conditionmetric-xopab9on47phx",namespace="ns-conditionmetric-zvhcfd973c3yg",status="True"} 1
redhat_appstudio_remotesecret_status_condition{condition="Deployed",name="test-conditionmetric-xopab9on47phx",namespace="ns-conditionmetric-zvhcfd973c3yg",status="False"} 1
```

3.  Create RS with data and with one target
```yaml
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: $OBJ_NAME
spec:
  secret:
    generateName: secret-from-remote-
  targets:
    - namespace: $TARGET_NS_NAMESPACE
stringData:
  a: b
  c: d
```
expected metric
```
redhat_appstudio_remotesecret_status_condition{condition="DataObtained",name="test-conditionmetric-g0pwha42nnp3e",namespace="ns-conditionmetric-lwzsskb9r1nor",status="True"} 1
redhat_appstudio_remotesecret_status_condition{condition="Deployed",name="test-conditionmetric-g0pwha42nnp3e",namespace="ns-conditionmetric-lwzsskb9r1nor",status="True"} 1
```

4. Two-stage upload
```yaml
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: $OBJ_NAME
spec:
  secret:
    generateName: some-secret-
    labels:
      a: b
    annotations:
      c: d
----
apiVersion: v1
kind: Secret
metadata:
  name: ${OBJ_NAME}-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: $OBJ_NAME
stringData:
  a: b
  c: d
```
expected metric
```
redhat_appstudio_remotesecret_status_condition{condition="DataObtained",name="test-conditionmetric-e85919rsghkca",namespace="ns-conditionmetric-mtfhqhr7a4b5r",status="False"} 0
redhat_appstudio_remotesecret_status_condition{condition="DataObtained",name="test-conditionmetric-e85919rsghkca",namespace="ns-conditionmetric-mtfhqhr7a4b5r",status="True"} 1
redhat_appstudio_remotesecret_status_condition{condition="Deployed",name="test-conditionmetric-e85919rsghkca",namespace="ns-conditionmetric-mtfhqhr7a4b5r",status="False"} 1

```

5. With data and two targets, one failed
```yaml
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: ${OBJ_NAME}
spec:
  secret:
    generateName: some-secret-
  targets:
    - namespace: ${TARGET_NAMESPACE}
    - namespace: non-existing-namespace
stringData:
  key: value
```
expected metric
```
redhat_appstudio_remotesecret_status_condition{condition="DataObtained",name="test-conditionmetric-ydu4covpx2e4i",namespace="ns-conditionmetric-72kglutn2lvo9",status="True"} 1
redhat_appstudio_remotesecret_status_condition{condition="Deployed",name="test-conditionmetric-ydu4covpx2e4i",namespace="ns-conditionmetric-72kglutn2lvo9",status="False"} 1

```
